Caused by: android.database.sqlite.SQLiteException

The GR****OUP BY no longer works on android Q.

```
Caused by: android.database.sqlite.SQLiteException: near "GROUP": syntax error (code 1 SQLITE_ERROR): , while compiling: SELECT bucket_id, bucket_display_name, _id FROM images WHERE ((is_pending=0) AND (is_trashed=0) AND (volume_name IN ( 'external_primary' ))) AND ((1) GROUP BY 1,(2)) ORDER BY MAX(datetaken) DESC
        at android.database.DatabaseUtils.readExceptionFromParcel(DatabaseUtils.java:184)
        at android.database.DatabaseUtils.readExceptionFromParcel(DatabaseUtils.java:140)
        at android.content.ContentProviderProxy.query(ContentProviderNative.java:423)
        at android.content.ContentResolver.query(ContentResolver.java:944)
        at android.content.ContentResolver.query(ContentResolver.java:880)
        at androidx.core.content.ContentResolverCompat.query(ContentResolverCompat.java:81)
        at androidx.loader.content.CursorLoader.loadInBackground(CursorLoader.java:63)
        at com.laevatein.internal.loader.AlbumLoader.loadInBackground(AlbumLoader.java:47)
        at com.laevatein.internal.loader.AlbumLoader.loadInBackground(AlbumLoader.java:34)
        at androidx.loader.content.AsyncTaskLoader.onLoadInBackground(AsyncTaskLoader.java:307)
        at androidx.loader.content.AsyncTaskLoader$LoadTask.doInBackground(AsyncTaskLoader.java:60)
        at androidx.loader.content.AsyncTaskLoader$LoadTask.doInBackground(AsyncTaskLoader.java:48)
        at androidx.loader.content.ModernAsyncTask$2.call(ModernAsyncTask.java:141)
        at java.util.concurrent.FutureTask.run(FutureTask.java:266)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1167) 
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:641) 
        at java.lang.Thread.run(Thread.java:919) 
```

ref: https://stackoverflow.com/questions/56823336/query-mediastore-on-android-q

--- a/.travis.yml
+++ b/.travis.yml
@@ -2,7 +2,7 @@ language: android
 jdk: oraclejdk8
 env:
  global:
-    - ANDROID_SDK=android-28
+    - ANDROID_SDK=android-29
     - ANDROID_BUILD_TOOLS_VERSION=28.0.3
     - ANDROID_ABI=armeabi-v7a
     - ANDROID_EMULATOR_SDK=android-22

--- a/laevatein/build.gradle
+++ b/laevatein/build.gradle
@@ -1,11 +1,11 @@
 apply plugin: 'com.android.library'
 
 android {
-    compileSdkVersion 28
+    compileSdkVersion 29
 
     defaultConfig {
         minSdkVersion 14
-        targetSdkVersion 28
+        targetSdkVersion 29
 
         testInstrumentationRunner "androidx.test.runner.AndroidJUnitRunner"
     }

--- a/laevatein/src/main/java/com/laevatein/internal/loader/AlbumLoader.java
+++ b/laevatein/src/main/java/com/laevatein/internal/loader/AlbumLoader.java
@@ -24,8 +24,8 @@ import androidx.loader.content.CursorLoader;
 
 import com.laevatein.internal.entity.Album;
 
-import java.util.HashSet;
-import java.util.Set;
+import java.util.LinkedHashMap;
+import java.util.Map;
 
 /**
  * Wrapper for {@link CursorLoader} to merge custom cursors.
@@ -37,7 +37,7 @@ import java.util.Set;
 public class AlbumLoader extends CursorLoader {
     public static final String TAG = AlbumLoader.class.getSimpleName();
     private static final String[] PROJECTION = { MediaStore.Images.Media.BUCKET_ID, MediaStore.Images.Media.BUCKET_DISPLAY_NAME, MediaStore.Images.Media._ID };
-    private static final String BUCKET_ORDER_BY = MediaStore.Images.Media.DATE_MODIFIED + " desc";
+    private static final String BUCKET_ORDER_BY = "datetaken DESC";
     private static final String MEDIA_ID_DUMMY = String.valueOf(-1);
 
     public AlbumLoader(Context context) {
@@ -48,53 +48,40 @@ public class AlbumLoader extends CursorLoader {
     public Cursor loadInBackground() {
         Cursor albums = super.loadInBackground();
 
-        Set<BucketEntry> buckets = new HashSet<>();
+        Map<String, BucketEntry> buckets = new LinkedHashMap<>();
         if (albums != null) {
             try {
                 while (albums.moveToNext()) {
-                    buckets.add(new BucketEntry(
-                            albums.getString(albums.getColumnIndex(MediaStore.Images.Media.BUCKET_ID)),
-                            albums.getString(albums.getColumnIndex(MediaStore.Images.Media.BUCKET_DISPLAY_NAME)),
-                            albums.getString(albums.getColumnIndex(MediaStore.Images.Media._ID))
-                    ));
+                    String bucketId = albums.getString(albums.getColumnIndex(MediaStore.Images.Media.BUCKET_ID));
+                    if (!buckets.containsKey(bucketId)) {
+                        buckets.put(bucketId, new BucketEntry(
+                                albums.getString(albums.getColumnIndex(MediaStore.Images.Media.BUCKET_DISPLAY_NAME)),
+                                albums.getString(albums.getColumnIndex(MediaStore.Images.Media._ID))
+                        ));
+                    }
                 }
             } finally {
                 albums.close();
             }
         }
+
         MatrixCursor allAlbum = new MatrixCursor(PROJECTION);
-        allAlbum.addRow(new String[] {Album.ALBUM_ID_ALL, Album.ALBUM_NAME_ALL, MEDIA_ID_DUMMY});
-        for (BucketEntry bucket : buckets) {
-            allAlbum.addRow(new String[] {bucket.bucketId, bucket.displayName, bucket.id});
+        allAlbum.addRow(new String[]{Album.ALBUM_ID_ALL, Album.ALBUM_NAME_ALL, MEDIA_ID_DUMMY});
+        for (Map.Entry<String, BucketEntry> entry : buckets.entrySet()) {
+            BucketEntry bucket = entry.getValue();
+            allAlbum.addRow(new String[]{entry.getKey(), bucket.displayName, bucket.mediaId});
         }
 
         return allAlbum;
     }
 
     private static class BucketEntry {
-        private final String bucketId;
         private final String displayName;
-        private final String id;
+        private final String mediaId;
 
-        public BucketEntry(String bucketId, String displayName, String id) {
-            this.bucketId = bucketId;
+        public BucketEntry(String displayName, String mediaId) {
             this.displayName = displayName;
-            this.id = id;
-        }
-
-        @Override
-        public boolean equals(Object o) {
-            if (this == o) return true;
-            if (o == null || getClass() != o.getClass()) return false;
-
-            BucketEntry bucket = (BucketEntry) o;
-
-            return bucketId.equals(bucket.bucketId);
-        }
-
-        @Override
-        public int hashCode() {
-            return bucketId.hashCode();
+            this.mediaId = mediaId;
         }
     }
 }

--- a/laevatein/src/main/java/com/laevatein/internal/loader/AlbumLoader.java
+++ b/laevatein/src/main/java/com/laevatein/internal/loader/AlbumLoader.java
@@ -15,14 +15,17 @@
  */
 package com.laevatein.internal.loader;
 
-import com.laevatein.internal.entity.Album;
-
 import android.content.Context;
 import android.database.Cursor;
 import android.database.MatrixCursor;
-import android.database.MergeCursor;
 import android.provider.MediaStore;
+
 import androidx.loader.content.CursorLoader;
+
+import com.laevatein.internal.entity.Album;
+
+import java.util.HashSet;
+import java.util.Set;
 
 /**
  * Wrapper for {@link CursorLoader} to merge custom cursors.
@@ -34,20 +37,64 @@ import androidx.loader.content.CursorLoader;
 public class AlbumLoader extends CursorLoader {
     public static final String TAG = AlbumLoader.class.getSimpleName();
     private static final String[] PROJECTION = { MediaStore.Images.Media.BUCKET_ID, MediaStore.Images.Media.BUCKET_DISPLAY_NAME, MediaStore.Images.Media._ID };
-    private static final String BUCKET_GROUP_BY = "1) GROUP BY 1,(2";
-    private static final String BUCKET_ORDER_BY = "MAX(datetaken) DESC";
+    private static final String BUCKET_ORDER_BY = MediaStore.Images.Media.DATE_MODIFIED + " desc";
     private static final String MEDIA_ID_DUMMY = String.valueOf(-1);
 
     public AlbumLoader(Context context) {
-        super(context, MediaStore.Images.Media.EXTERNAL_CONTENT_URI, PROJECTION, BUCKET_GROUP_BY, null, BUCKET_ORDER_BY);
+        super(context, MediaStore.Images.Media.EXTERNAL_CONTENT_URI, PROJECTION, null, null, BUCKET_ORDER_BY);
     }
 
     @Override
     public Cursor loadInBackground() {
         Cursor albums = super.loadInBackground();
+
+        Set<BucketEntry> buckets = new HashSet<>();
+        if (albums != null) {
+            try {
+                while (albums.moveToNext()) {
+                    buckets.add(new BucketEntry(
+                            albums.getString(albums.getColumnIndex(MediaStore.Images.Media.BUCKET_ID)),
+                            albums.getString(albums.getColumnIndex(MediaStore.Images.Media.BUCKET_DISPLAY_NAME)),
+                            albums.getString(albums.getColumnIndex(MediaStore.Images.Media._ID))
+                    ));
+                }
+            } finally {
+                albums.close();
+            }
+        }
         MatrixCursor allAlbum = new MatrixCursor(PROJECTION);
         allAlbum.addRow(new String[] {Album.ALBUM_ID_ALL, Album.ALBUM_NAME_ALL, MEDIA_ID_DUMMY});
+        for (BucketEntry bucket : buckets) {
+            allAlbum.addRow(new String[] {bucket.bucketId, bucket.displayName, bucket.id});
+        }
 
-        return new MergeCursor(new Cursor[]{ allAlbum, albums });
+        return allAlbum;
+    }
+
+    private static class BucketEntry {
+        private final String bucketId;
+        private final String displayName;
+        private final String id;
+
+        public BucketEntry(String bucketId, String displayName, String id) {
+            this.bucketId = bucketId;
+            this.displayName = displayName;
+            this.id = id;
+        }
+
+        @Override
+        public boolean equals(Object o) {
+            if (this == o) return true;
+            if (o == null || getClass() != o.getClass()) return false;
+
+            BucketEntry bucket = (BucketEntry) o;
+
+            return bucketId.equals(bucket.bucketId);
+        }
+
+        @Override
+        public int hashCode() {
+            return bucketId.hashCode();
+        }
     }
 }

--- a/sample/build.gradle
+++ b/sample/build.gradle
@@ -1,11 +1,11 @@
 apply plugin: 'com.android.application'
 
 android {
-    compileSdkVersion 28
+    compileSdkVersion 29
 
     defaultConfig {
         minSdkVersion 14
-        targetSdkVersion 26
+        targetSdkVersion 29
         versionCode 1
         versionName "1.0"
     }

